### PR TITLE
Fix for infinite materials from the autolathe.

### PR DESCRIPTION
**Pull request by WarlockD**
## About The Pull Request
Its about issue #52852.  It took a bit to find honestly.  Two issues are happening.

First, the autolathe does NOT set custom_materials to the materials being used.  The material_container uses this list to figure out how to disassemble the object so otherwise it uses the default setting for the object.  This has been fixed.

Second, ammo boxes self set itself the right material cost, but never set the bullets.   So once it's plowed through the recycler, the contents, are disassembled at full cost each as well as the box, witch is ALSO set to ALL the bullet costs.  So you would always get double your money with each box.  THIS has been fixed...I think.

Someone check my work though.  I think I got though all the use cases but I haven't checked if all obj/items have issues.

## Why It's Good For The Game

Unlimited materials is op doog.

## Changelog
:cl:
fix: No more unlimited materials from the autolathe
/:cl:

### CODE REVIEW

**Comment by WarlockD** [comment]
Oops, forgot to remove the custom materials from the box when I moved them to the bullets meh

**Comment by Winterous** [comment]
Rip matter replicator.
Funny thing is I'm certain this didn't work a few years ago.

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm [hunk 1]
```diff
@@ -35,18 +35,42 @@
...
 
+/**
+  * Called to recalculate the matinal cost of a bullet
+  *
+  * Should only be called when the custom_materials is changed
+  * so usually on create
+  *
+  */
+/obj/item/ammo_box/proc/_refresh_bullet_cost()
+	var/ammo_count = stored_ammo?.len || max_ammo
```
Comment (Rohesie) [inline code/modules/projectiles/boxes_magazines/_box_magazine.dm:53]:
```suggestion
	var/ammo_count = length(stored_ammo) || max_ammo
```

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm [hunk 2]
```diff
@@ -35,18 +35,42 @@
...
+	bullet_cost = list()
+	base_cost = list()
+	for (var/material in custom_materials)
+		var/material_amount = custom_materials[material]
+		base_cost[material] = (material_amount * 0.10)
+		material_amount *= 0.90 // 10% for the container
+		material_amount /= ammo_count
+		bullet_cost[material] = material_amount
+	if(length(stored_ammo))
+		for(var/atom/A in stored_ammo) // hard set all the bullets to this cost
```
Comment (Rohesie) [inline code/modules/projectiles/boxes_magazines/_box_magazine.dm:63]: Can `stored_ammo` hold anything that's not an `/atom`? If not, then there's no need to type check here.

**Comment by Qustinnus** [comment]
`First, the autolathe does NOT set custom_materials to the materials being used. The material_container uses this list to figure out how to disassemble the object so otherwise it uses the default setting for the object. This has been fixed.`

this is intended, the default settings should be the same as the recipe costs. calling set_custom_materials on everything another time is not really a solution.

File: code/game/machinery/autolathe.dm [hunk 3]
```diff
@@ -228,14 +228,21 @@
...
 
 			if(length(picked_materials))
 				new_item.set_custom_materials(picked_materials, 1 / multiplier) //Ensure we get the non multiplied amount
 				for(var/x in picked_materials)
 					var/datum/material/M = x
 					if(!istype(M, /datum/material/glass) && !istype(M, /datum/material/iron))
 						user.client.give_award(/datum/award/achievement/misc/getting_an_upgrade, user)
+			else
+				// we STILL need to set custom materials or we get unlimited autolathe works happening
+				var/old_material_flags = new_item.material_flags
```
Comment (Qustinnus) [inline code/game/machinery/autolathe.dm:240]: this entire "set the flag then set the custom materials then unset the flags" makes no sense. Why would you temporary make materials not have effects then unset it? that's just going to make it confusing to work with.

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm (hunk 2 above)
Comment (WarlockD) [reply code/modules/projectiles/boxes_magazines/_box_magazine.dm:63]: It shouldn't.  Is it slower than casting it though?  I thought all this does is just loop through a list and skip non typed?  I still need to cast it as an atom to set its custom_material cost

File: code/game/machinery/autolathe.dm (hunk 3 above)
Comment (WarlockD) [reply code/game/machinery/autolathe.dm:240]: I know, but...humm

I was going to argue the point that some items change their specs depending on the materials being used, but now that I think about it, it doesn't matter does it.  In fact, we don't even need to have two separate "material" lists go to this function.  The flags are just there so the user can choose what materials are used.  Let me take a look at it again.

**Comment by WarlockD** [comment]
> `First, the autolathe does NOT set custom_materials to the materials being used. The material_container uses this list to figure out how to disassemble the object so otherwise it uses the default setting for the object. This has been fixed.`
> 
> this is intended, the default settings should be the same as the recipe costs. calling set_custom_materials on everything another time is not really a solution.

Humm, so can I get away with just setting the var custom_materials?  I am going to take more of a look at it

**Comment by WarlockD** [comment]
Ok did some more cleaning and changed how custom_matierals is applied.  Your right as you don't have to save the flag.

File: code/modules/projectiles/boxes_magazines/_box_magazine.dm (hunk 2 above)
Comment (Rohesie) [reply code/modules/projectiles/boxes_magazines/_box_magazine.dm:63]: Prefer to use something like this:
```
for(var/a in stored_ammo)
  var/atom/ammo = a
```
This has two advantages:
1) It is faster, as it doesn't have to typecheck every loop. Since we know the contents of the list there's no need for checking.
2) In case our list gets polluted by `nulls`, more often than not for improperly handled reference cleaning on deletion, it will runtime and warn us about it. Nulls in list are pain, and we should know about them.

File: code/game/machinery/autolathe.dm [hunk 4]
```diff
@@ -228,15 +244,16 @@
 	else
 		for(var/i=1, i<=multiplier, i++)
 			var/obj/item/new_item = new being_built.build_path(A)
-			new_item.autolathe_crafted(src)
 
-			if(length(picked_materials))
```
Comment (Qustinnus) [inline code/game/machinery/autolathe.dm:233]: this check was there for a reason, why did you remove it?

File: code/game/machinery/autolathe.dm [hunk 5]
```diff
@@ -228,15 +244,16 @@
...
-					var/datum/material/M = x
-					if(!istype(M, /datum/material/glass) && !istype(M, /datum/material/iron))
-						user.client.give_award(/datum/award/achievement/misc/getting_an_upgrade, user)
+			if(!(new_item.material_flags & MATERIAL_NO_EFFECTS))
+				new_item.set_custom_materials(materials_used) // the object has effects that change the object depending on the material.
+				if(!isnull(user))
+					// Get an achievement for making a gold toolbox!
+					user.client.give_award(/datum/award/achievement/misc/getting_an_upgrade, user)
+			else
+				new_item.custom_materials = materials_used // it has no effects, so just set the amount of materials being used to make it
```
Comment (Qustinnus) [inline code/game/machinery/autolathe.dm:254]: this is bad, the setter exists for a reason.

File: code/game/machinery/autolathe.dm [hunk 6]
```diff
@@ -228,15 +244,16 @@
...
 
-			if(length(picked_materials))
-				new_item.set_custom_materials(picked_materials, 1 / multiplier) //Ensure we get the non multiplied amount
-				for(var/x in picked_materials)
-					var/datum/material/M = x
-					if(!istype(M, /datum/material/glass) && !istype(M, /datum/material/iron))
-						user.client.give_award(/datum/award/achievement/misc/getting_an_upgrade, user)
+			if(!(new_item.material_flags & MATERIAL_NO_EFFECTS))
+				new_item.set_custom_materials(materials_used) // the object has effects that change the object depending on the material.
+				if(!isnull(user))
```
Comment (Qustinnus) [inline code/game/machinery/autolathe.dm:250]: just do user?. infront of the client.give_award() instead of doing an isnull check

File: code/game/machinery/autolathe.dm [hunk 7]
```diff
@@ -228,15 +244,16 @@
...
-			new_item.autolathe_crafted(src)
 
-			if(length(picked_materials))
-				new_item.set_custom_materials(picked_materials, 1 / multiplier) //Ensure we get the non multiplied amount
-				for(var/x in picked_materials)
-					var/datum/material/M = x
-					if(!istype(M, /datum/material/glass) && !istype(M, /datum/material/iron))
-						user.client.give_award(/datum/award/achievement/misc/getting_an_upgrade, user)
+			if(!(new_item.material_flags & MATERIAL_NO_EFFECTS))
+				new_item.set_custom_materials(materials_used) // the object has effects that change the object depending on the material.
```
Comment (Qustinnus) [inline code/game/machinery/autolathe.dm:249]: It should only run this if it uses custom mats, which is why picked_mats existed.

File: code/game/machinery/autolathe.dm [hunk 8]
```diff
@@ -228,15 +244,16 @@
...
+				if(!isnull(user))
+					// Get an achievement for making a gold toolbox!
+					user.client.give_award(/datum/award/achievement/misc/getting_an_upgrade, user)
```
Comment (Qustinnus) [inline code/game/machinery/autolathe.dm:252]: you get an achievement for making any toolbox like this.

Comment (WarlockD) [reply code/game/machinery/autolathe.dm:252]: Well yea.  It was like that originally for some reason.

File: code/game/machinery/autolathe.dm (hunk 6 above)
Comment (WarlockD) [reply code/game/machinery/autolathe.dm:250]: I wanted to save a loop.  It checks it in the previous call and sets the user null or not depending.  Its the only point where user is used in this fucntion.

File: code/game/machinery/autolathe.dm (hunk 8 above)
Comment (WarlockD) [reply code/game/machinery/autolathe.dm:252]: But it checks if it was not made out of metal or glass in the previous call.  Again, not saying its not a stupid achievement, just saying it still tests the same thing.

**Comment by WarlockD** [comment]
Meh I will work on this latter or not I don't know
